stakerservice/rpc_func.go

```diff
@@ -0,0 +1,580 @@
+package stakerservice
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"reflect"
+	"regexp"
+	"sort"
+	"strings"
+
+	cmtjson "github.com/cometbft/cometbft/libs/json"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cometbft/cometbft/rpc/jsonrpc/server"
+	types "github.com/cometbft/cometbft/rpc/jsonrpc/types"
+)
+
+// https://github.com/cometbft/cometbft/blob/38a4caeac0551c188af8a3e209e48380cd514e2d/rpc/jsonrpc/server/rpc_func.go#L1
+// This file was mostly coppied from cometbft just to add a middleware in the RPC routes.
+// The middleware used was to add basic auth to routes with user and password
+
+var reInt = regexp.MustCompile(`^-?[0-9]+$`)
+
+// RPCFunc contains the introspected type information for a function
+type RPCFunc struct {
+	f              reflect.Value          // underlying rpc function
+	args           []reflect.Type         // type of each function arg
+	returns        []reflect.Type         // type of each return arg
+	argNames       []string               // name of each argument
+	cacheable      bool                   // enable cache control
+	ws             bool                   // enable websocket communication
+	noCacheDefArgs map[string]interface{} // a lookup table of args that, if not supplied or are set to default values, cause us to not cache
+}
+
+type Option func(*RPCFunc)
+
+// RegisterRPCFuncs adds a route for each function in the funcMap, as well as
+// general jsonrpc and websocket handlers for all functions. "result" is the
+// interface on which the result objects are registered, and is popualted with
+// every RPCResponse
+func RegisterRPCFuncs(mux *http.ServeMux, funcMap map[string]*RPCFunc, logger log.Logger, middleware func(http.HandlerFunc) http.HandlerFunc) {
+	// HTTP endpoints
+	for funcName, rpcFunc := range funcMap {
+		rpcHandler := makeHTTPHandler(rpcFunc, logger)
+		mux.HandleFunc("/"+funcName, middleware(rpcHandler))
+	}
+
+	// JSONRPC endpoints
+	mux.HandleFunc("/", handleInvalidJSONRPCPaths(makeJSONRPCHandler(funcMap, logger)))
```

Comment:
shouldn't we also use authethication here ?

Reply:
we can authenticate this route as well, but basically it only shows the endpoints

Reply:
I think it will only show endpoints if we call it with empty payload, if we call is with proper json rpc body it will do rpc call to correct function

Reply:
> I think it will only show endpoints if we call it with empty payload, if we call is with proper json rpc body it will do rpc call to correct function

You were correct, indeed it needed the middleware to authenticate the "/" path as well